insonmnia/hardware/hardware.go

```diff
@@ -68,6 +80,94 @@ func (h *Hardware) SetNetworkIncoming(IPs []string) {
 	}
 }
 
+func (h *Hardware) AskPlanResources() *sonm.AskPlanResources {
+	result := sonm.NewEmptyAskPlanResources()
+	result.CPU.CorePercents = uint64(h.CPU.GetDevice().GetCores()) * 100
+	result.RAM.Size.Bytes = h.RAM.Device.Available
+	result.Storage.Size.Bytes = h.Storage.GetDevice().GetBytesAvailable()
+	for _, gpu := range h.GPU {
+		result.GPU.Hashes = append(result.GPU.Hashes, gpu.Device.Hash)
+	}
+	result.Network.Outbound = h.Network.Outbound
+	result.Network.Overlay = h.Network.Overlay
+	result.Network.Incoming = h.Network.Incoming
+	//TODO: Make network device use DataSizeRate
+	result.Network.ThroughputIn.BitsPerSecond = h.Network.GetIn()
+	result.Network.ThroughputOut.BitsPerSecond = h.Network.GetOut()
+	return result
+}
+
+func insertBench(to []uint64, bench *sonm.Benchmark, proportion float64) ([]uint64, error) {
+	for len(to) <= int(bench.ID) {
+		to = append(to, uint64(0))
+	}
+	if bench.SplittingAlgorithm == sonm.SplittingAlgorithm_NONE {
+		if to[bench.ID] != 0 {
+			return nil, fmt.Errorf("duplicate benchmark with id %d and type none", bench.ID)
+		}
+		to[bench.ID] = bench.GetResult()
+	} else if bench.SplittingAlgorithm == sonm.SplittingAlgorithm_PROPORTIONAL {
+		to[bench.ID] += uint64(float64(bench.Result) * proportion)
+	} else if bench.SplittingAlgorithm == sonm.SplittingAlgorithm_MAX {
+		if bench.Result > to[bench.ID] {
+			to[bench.ID] = bench.Result
+		}
+	} else if bench.SplittingAlgorithm == sonm.SplittingAlgorithm_MIN {
+		if bench.Result < to[bench.ID] {
+			to[bench.ID] = bench.Result
+		}
+	}
+	return to, nil
+}
+
+func (h *Hardware) ResourcesToBenchmarks(resources *sonm.AskPlanResources) (*sonm.Benchmarks, error) {
+	if !resources.GPU.Normalized() {
+		return nil, errors.New("passed resources are not normalized, call resources.GPU.Normalize(hardware) first")
+	}
+	var err error
+	benchmarks := make([]uint64, sonm.MinNumBenchmarks)
+
+	proportions := []float64{}
+	hwBenches := []map[uint64]*sonm.Benchmark{}
+	for _, hash := range resources.GetGPU().GetHashes() {
+		found := false
+		for _, gpu := range h.GPU {
+			if gpu.GetDevice().GetHash() == hash {
+				hwBenches = append(hwBenches, gpu.Benchmarks)
+				proportions = append(proportions, 1.0)
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("unknown hash in passed resources - %s", hash)
+		}
+	}
+
+	hwBenches = append(hwBenches,
```

Comment:
I don't like this. Looks very error prone.

Reply:
alternative is tons of err!=nil checking